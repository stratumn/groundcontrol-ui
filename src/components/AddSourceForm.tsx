
// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React, { Component, Fragment } from "react";
import {
  Button,
  DropdownProps,
  Form,
  InputProps,
} from "semantic-ui-react";

export enum SourceType { Directory = "directory", Git = "git" }

export interface IProps {
  type: SourceType;
  directory: string;
  repository: string;
  reference: string;
  onChange: (values: IProps) => any;
  onSubmit: (values: IProps) => any;
}

export default class AddSourceForm extends Component<IProps> {

  public render() {
    const { type, directory, repository, reference } = this.props;
    const options = [
      { key: SourceType.Directory, text: "Directory", value: SourceType.Directory },
      { key: SourceType.Git, text: "Git", value: SourceType.Git },
    ];
    const disabled = type === SourceType.Directory && !directory ||
      type === SourceType.Git && !repository;

    let typeFields: JSX.Element;

    if (type === "directory") {
      typeFields = (
        <Form.Field>
          <label>Directory (absolute path)</label>
          <Form.Input
            name="directory"
            value={directory}
            onChange={this.handleChangeInput}
          />
        </Form.Field>
      );
    } else {
      typeFields = (
        <Fragment>
          <Form.Field>
            <label>Repository</label>
            <Form.Input
              name="repository"
              value={repository}
              onChange={this.handleChangeInput}
            />
          </Form.Field>
          <Form.Field>
            <label>Reference</label>
            <Form.Input
              name="reference"
              placeholder="refs/heads/master"
              value={reference}
              onChange={this.handleChangeInput}
            />
          </Form.Field>
        </Fragment>
      );
    }

    return (
      <Form onSubmit={this.handleSubmit}>
        <Form.Select
          label="Type"
          options={options}
          value={type}
          width="3"
          onChange={this.handleChangeType}
        />
        {typeFields}
        <Button
          type="submit"
          color="teal"
          icon="add"
          content="Add"
          disabled={disabled}
        />
      </Form>
    );
  }

  private handleSubmit = () => {
    this.props.onSubmit({ ...this.props });
  }

  private handleChangeType = (_: React.SyntheticEvent<HTMLElement>, { value }: DropdownProps) => {
    switch (value) {
    case "directory":
    case "git":
      this.props.onChange({ ...this.props, type: value as SourceType });
      break;
    }
  }

  private handleChangeInput = (_: React.SyntheticEvent<HTMLElement>, { name, value }: InputProps) => {
    const { onChange } = this.props;

    switch (name) {
    case "directory":
      onChange({ ...this.props, directory: value });
      break;
    case "repository":
      onChange({ ...this.props, repository: value });
      break;
    case "reference":
      onChange({ ...this.props, reference: value });
      break;
    }
  }

}
