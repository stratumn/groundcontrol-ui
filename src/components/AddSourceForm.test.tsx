// Copyright 2019 Stratumn
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { shallow } from "enzyme";
import React from "react";
import { mocked } from "ts-jest/utils";

import AddSourceForm, { SourceType } from "./AddSourceForm";

const props = {
  directory: "directory",
  onChange: jest.fn(),
  onSubmit: jest.fn(),
  reference: "reference",
  repository: "repository",
  type: SourceType.Directory,
};

beforeEach(() => {
  mocked(props.onChange).mockClear();
  mocked(props.onSubmit).mockClear();
});

describe("<AddSourceForm />", () => {

  it("renders correctly when a directory source type is selected", () => {
    const wrapper = shallow(
      <AddSourceForm
        {...props}
        type={SourceType.Directory}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("renders correctly when a Git source type is selected", () => {
    const wrapper = shallow(
      <AddSourceForm
        {...props}
        type={SourceType.Git}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("triggers onChange when a source type is selected", () => {
    const wrapper = shallow(
      <AddSourceForm
        {...props}
        type={SourceType.Directory}
      />,
    );
    wrapper.find("[label='Type']").simulate("change", null, {
      value: "git",
    });
    expect(props.onChange).toBeCalledTimes(1);
    expect(props.onChange).toBeCalledWith({...props, type: "git"});
  });

  it("triggers onChange when a value is changed", () => {
    const wrapper = shallow(
      <AddSourceForm
        {...props}
        type={SourceType.Directory}
      />,
    );
    wrapper.find("[name='directory']").simulate("change", null, {
      name: "directory",
      value: "new",
    });
    expect(props.onChange).toBeCalledTimes(1);
    expect(props.onChange).toBeCalledWith({...props, directory: "new"});
  });

  it("triggers onSubmit when the form is submitted", () => {
    const wrapper = shallow(
      <AddSourceForm
        {...props}
        type={SourceType.Directory}
      />,
    );
    wrapper.find("Form").simulate("submit");
    expect(props.onSubmit).toBeCalledTimes(1);
    expect(props.onSubmit).toBeCalledWith(props);
  });

});
