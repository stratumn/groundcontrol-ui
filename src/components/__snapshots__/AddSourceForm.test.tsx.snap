// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddSourceForm /> renders correctly when a Git source type is selected 1`] = `
<Form
  as="form"
  onSubmit={[Function]}
>
  <FormSelect
    as={[Function]}
    control={[Function]}
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "directory",
          "text": "Directory",
          "value": "directory",
        },
        Object {
          "key": "git",
          "text": "Git",
          "value": "git",
        },
      ]
    }
    value="git"
    width="3"
  />
  <FormField>
    <label>
      Repository
    </label>
    <FormInput
      as={[Function]}
      control={[Function]}
      name="repository"
      onChange={[Function]}
      value="repository"
    />
  </FormField>
  <FormField>
    <label>
      Reference
    </label>
    <FormInput
      as={[Function]}
      control={[Function]}
      name="reference"
      onChange={[Function]}
      placeholder="refs/heads/master"
      value="reference"
    />
  </FormField>
  <Button
    as="button"
    color="teal"
    content="Add"
    disabled={false}
    icon="add"
    type="submit"
  />
</Form>
`;

exports[`<AddSourceForm /> renders correctly when a directory source type is selected 1`] = `
<Form
  as="form"
  onSubmit={[Function]}
>
  <FormSelect
    as={[Function]}
    control={[Function]}
    label="Type"
    onChange={[Function]}
    options={
      Array [
        Object {
          "key": "directory",
          "text": "Directory",
          "value": "directory",
        },
        Object {
          "key": "git",
          "text": "Git",
          "value": "git",
        },
      ]
    }
    value="directory"
    width="3"
  />
  <FormField>
    <label>
      Directory (absolute path)
    </label>
    <FormInput
      as={[Function]}
      control={[Function]}
      name="directory"
      onChange={[Function]}
      value="directory"
    />
  </FormField>
  <Button
    as="button"
    color="teal"
    content="Add"
    disabled={false}
    icon="add"
    type="submit"
  />
</Form>
`;
